### PR TITLE
feat: migrate playground to flat config

### PR DESCRIPTION
**Pull request by snitin315**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

Migration of the online playground to use flat config system.

#### What changes did you make? (Give an overview)

- Update playground to use flat config.
- Removed environments.
- Supports existing links to the Playground by translating configurations to the new format.
- Updated download configuration functionality
   - ~If `sourceType` is `module` then the downloaded config is in the ESM format otherwise it's in the CommonJS format.~
   - Added an option to select config file format.
 
Screenshot 2024-01-26 at 4 38 58 PM

#### Related Issues

Fix https://github.com/eslint/eslint.org/issues/507

#### Is there anything you'd like reviewers to focus on?
Should we update the default value for `sourceType` to `module`?

### CODE REVIEW

File: src/playground/components/Configuration.js
Comment (nzakas) [inline src/playground/components/Configuration.js:132]: I don't think this is a safe assumption to make, at least, not based on the feedback we received about wanting `eslint.config.cjs` and `eslint.config.mjs` support. It's probably better to offer the choice of download format.

**Review by nzakas — CHANGES_REQUESTED** [review]
Thanks for doing this. Just left a few comments that I believe answer most of your questions.

File: src/playground/components/Configuration.js
Comment (snitin315) [reply src/playground/components/Configuration.js]: Updated, now users can choose the format.
Screenshot 2024-01-25 at 3 48 39 PM

File: src/playground/App.js
Comment (snitin315) [reply src/playground/App.js]: Yes, udpated.

File: src/playground/components/Configuration.js
Comment (kecrily) [reply src/playground/components/Configuration.js]: This option is too far away from the download button. Maybe we can make it closer to the download button and remove the text about formatting on the button. Turning it into two buttons would be nice.

Comment (snitin315) [reply src/playground/components/Configuration.js]: I've moved the selector closer to the button.

Screenshot 2024-01-26 at 4 38 58 PM

**Comment by nzakas** [comment]
This is looking really good! It appears there's a merge conflict, though.

**Comment by snitin315** [comment]
@nzakas I've fixed the conflicts.

**Review by nzakas — APPROVED** [review]
LGTM, thanks! Leaving open to give others the opportunity to review before merging.

File: src/playground/App.js
Comment (mdjermanovic) [inline src/playground/App.js:47]: When I open https://deploy-preview-509--new-eslint.netlify.app/play/ with a clean cache, recommended rules are not enabled, so it looks like this branch never happens. Maybe because `convertLegacyOptionsToFlatConfig` always returns an object?

File: src/playground/App.js
Comment (mdjermanovic) [inline src/playground/App.js:74]: Links from the docs site don't work. For example, this is with `#...` copied from the first example for the [`accessor-pairs`](https://eslint.org/docs/latest/rules/accessor-pairs#setwithoutget) rule:

https://deploy-preview-509--new-eslint.netlify.app/play/#eyJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsic291cmNlVHlwZSI6Im1vZHVsZSJ9fSwidGV4dCI6Ii8qZXNsaW50IGFjY2Vzc29yLXBhaXJzOiBcImVycm9yXCIqL1xuXG52YXIgbyA9IHtcbiAgICBzZXQgYSh2YWx1ZSkge1xuICAgICAgICB0aGlzLnZhbCA9IHZhbHVlO1xuICAgIH1cbn07XG5cblxudmFyIG8gPSB7ZDogMX07XG5PYmplY3QuZGVmaW5lUHJvcGVydHkobywgJ2MnLCB7XG4gICAgc2V0OiBmdW5jdGlvbih2YWx1ZSkge1xuICAgICAgICB0aGlzLnZhbCA9IHZhbHVlO1xuICAgIH1cbn0pOyJ9

```
playground.js:2 TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at rv (playground.js:2:2203255)
    at wo (playground.js:2:131194)
    at Cl (playground.js:2:188149)
    at bc (playground.js:2:177131)
    at Ec (playground.js:2:177059)
    at yc (playground.js:2:176922)
    at oc (playground.js:2:173729)
    at sc (playground.js:2:172284)
    at C (playground.js:2:279135)
```

When I try this locally (with the debug info), it seems that the error is thrown when trying to do `Object.keys(options.rules)`.

I _think_ this happens because this function always creates `rules` property, even if the value is `undefined`, so in `fillOptionsDefaults` it overwrites `rules: {}`.

Comment (snitin315) [reply src/playground/App.js:74]: I think this broke after I rebased on top of https://github.com/eslint/eslint.org/commit/44cfc74d2de96c9bcaa6c143b1a372fc938abb5f, I'll push a fix.

File: src/playground/App.js
Comment (snitin315) [reply src/playground/App.js:88]: Good Catch! Ideally, it would be great if we could set an `error.code = "ESLINT_VALIDATION_ERROR` or a specific exit code in case of validation errors. Right now I'm evaluating it based on the presence of`key "rules"` in error message.

File: src/playground/App.js
Comment (snitin315) [reply src/playground/App.js:49]: You're right. Pushed a fix.

**Comment by snitin315** [comment]
@mdjermanovic I've resolved your comments

File: src/playground/App.js
Comment (aladdin-add) [inline src/playground/App.js:41]: it will throw an error when `options.languageOptions` is `null/undef`?

File: src/playground/components/Configuration.js
Comment (aladdin-add) [inline src/playground/components/Configuration.js:187]: Set to "latest" if it's invalid?
```suggestion
                                    ECMAVersionsOptions.filter(ecmaVersion => ecmaVersion.value === (options.languageOptions.ecmaVersion)) || "latest"
```

File: src/playground/App.js
Comment (aladdin-add) [inline src/playground/App.js:63]: should also check `env` - it's away in the new config.

File: src/playground/components/Configuration.js
Comment (aladdin-add) [inline src/playground/components/Configuration.js:203]: same here?

File: src/playground/components/Configuration.js
Comment (mdjermanovic) [inline src/playground/components/Configuration.js:143]: Same question here.

**Comment by nzakas** [comment]
@snitin315 are you still working on this?

**Comment by mdjermanovic** [comment]
Since @snitin315 is away, I'll take over this branch to finish this up.

File: src/playground/App.js
Comment (mdjermanovic) [reply src/playground/App.js:41]: `{ ...undefined }` doesn't throw, it's just a no-op.

**Comment by mdjermanovic** [comment]
> > If we want to rely on Linter's defaults, then I think the dropdowns should have an empty item that would be selected by default. Otherwise, we could set defaults in the Playground's config as it used to be before this change, which I think is fine.
> 
> Rather than an empty item, maybe "(default)" for clarity? I'd prefer not to always include `sourceType` and `ecmaVersion` in the config, as we should be encouraging people to use the defaults unless they have a specific need.

I've added a "(default)" item in the two lists.

File: src/playground/App.js
Comment (mdjermanovic) [reply src/playground/App.js]: This is now refactored a bit, please check the new version.

File: src/playground/components/Configuration.js
Comment (mdjermanovic) [reply src/playground/components/Configuration.js]: https://github.com/eslint/eslint.org/pull/509#discussion_r1541507365

**Comment by mdjermanovic** [comment]
Ready for re-review.

**Comment by nzakas** [comment]
I tested a few different combinations and verified that the `ecmaVersion` and `sourceType` fields do not appear in config files when the default values are used.

I did notice that `parserOptions.ecmaFeatures` is always present, even when I didn't add any ecmaFeatures. Is that expected?

**Comment by mdjermanovic** [comment]
> I did notice that `parserOptions.ecmaFeatures` is always present, even when I didn't add any ecmaFeatures. Is that expected?

It works that way in the current https://eslint.org/play/, and in the old https://archive.eslint.org/demo too.

I'm not sure if this was intentional. It could be that the idea was to always create an object, even if empty, so that the user could easily add to it later.

Either way, I think it's fine to remove `ecmaFeatures` and its parent `parserOptions` object altogether if no ECMA Features were selected.

Furthermore, we could also remove `languageOptions` if there are no ECMA Features selected and ECMA Version & Souce Type are both "(default)"?

File: src/playground/App.js
Comment (aladdin-add) [inline src/playground/App.js:19]: we had reached a consensus to https://github.com/eslint/eslint/issues/18103#issuecomment-1941230245.

Comment (mdjermanovic) [reply src/playground/App.js:19]: Agreed, but that will need some time to implement, so I think we can leave it for a follow-up. This will work until at least v10 :)

Comment (aladdin-add) [reply src/playground/App.js:19]: haha, it also makes sense.

**Review by aladdin-add — APPROVED** [review]
LGTM, pending @nzakas 's review.

**Comment by nzakas** [comment]
> > I did notice that `parserOptions.ecmaFeatures` is always present, even when I didn't add any ecmaFeatures. Is that expected?
> 
> It works that way in the current https://eslint.org/play/, and in the old https://archive.eslint.org/demo too.
> 
> I'm not sure if this was intentional. It could be that the idea was to always create an object, even if empty, so that the user could easily add to it later.
> 
> Either way, I think it's fine to remove `ecmaFeatures` and its parent `parserOptions` object altogether if no ECMA Features were selected.
> 
> Furthermore, we could also remove `languageOptions` if there are no ECMA Features selected and ECMA Version & Source Type are both "(default)"?

Yes, I think we don't need any object that doesn't have any overrides in it.

**Comment by mdjermanovic** [comment]
> > > I did notice that `parserOptions.ecmaFeatures` is always present, even when I didn't add any ecmaFeatures. Is that expected?
> > 
> > 
> > It works that way in the current https://eslint.org/play/, and in the old https://archive.eslint.org/demo too.
> > I'm not sure if this was intentional. It could be that the idea was to always create an object, even if empty, so that the user could easily add to it later.
> > Either way, I think it's fine to remove `ecmaFeatures` and its parent `parserOptions` object altogether if no ECMA Features were selected.
> > Furthermore, we could also remove `languageOptions` if there are no ECMA Features selected and ECMA Version & Source Type are both "(default)"?
> 
> Yes, I think we don't need any object that doesn't have any overrides in it.

Done in https://github.com/eslint/eslint.org/pull/509/commits/ae67deb7755ec279cdb4578394646b9d48e21df8.

I noticed that in some cases Linter modifies passed configs (for example, [here](https://github.com/eslint/eslint/blob/9b7bd3be066ac1f72fa35c4d31a1b178c7e2b683/lib/linter/linter.js#L1686)), which causes confusing behavior in the Playground, so I've updated code to pass a cloned config in https://github.com/eslint/eslint.org/pull/509/commits/f5ecbe7fa30d4bde313af4ee3a46568ca3f59435.

I also added ecmaVersion normalization in https://github.com/eslint/eslint.org/pull/509/commits/e4ae6cd7b264ab83241682cd2a58a03b36374616, as per the discussion in https://github.com/eslint/eslint/pull/18241.

**Review by nzakas — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**
